### PR TITLE
cpu/sam0_common: use generic exti_config

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Creating an `exti_config` array for a new MCU manually is tedious and error prone.
Luckiely all information is already availiable in the vendor files.

Credit for this discovery & method goes to @Sizurka

The file was generated with

```C

int main(void) {
        puts("static const int8_t exti_config[PORT_GROUPS][32] = {");

        for (unsigned port = 1; port < 5; ++port) {
                printf("#if PORT_GROUPS >= %d\n{\n", port);
                for (unsigned pin = 0; pin < 32; ++pin) {
                        printf("#ifdef PIN_P%c%02uA_EIC_EXTINT_NUM\n", '@' + port, pin);
                        printf("    PIN_P%c%02uA_EIC_EXTINT_NUM,\n", '@' + port, pin);
                        printf("#else\n    -1,\n#endif\n");
                }
                printf("},\n#endif\n\n");
        }

        puts("};");

        return 0;
}
```

No changes in generated code are expected, but this makes adding new members
of the sam0 CPU families much easier.

### Testing procedure

External interrupts should work as before, in fact the same binary should be generated.

If your board has a module connected that uses GPIO interrupts (e.g. the radio in `samr21-xpro`), they should still continue to work.
If your board has a button, you can try `tests/periph_gpio`.

e.g. to configure an interrupt for the falling flank of the user button (`PB31`) on `same54-xpro` run

```
init_int 1 31 0
```

Pressing the button should trigger a message on the screen

*tested on `samr21-xpro`,  `saml21-xpro` and `same54-xpro`*

### Issues/PRs references

For samd21 and samr21 updated vendor files are needed as those defines were added in a later version than the one currently used by RIOT.

~~#13313 for samd21~~
~~#13350 for saml21~~
~~#13311 for samr21~~
~~#13402 for samr30~~

### CODE REVIEW

**Comment by dylad** [comment]
I like this solution, this is really good !
Thanks for keeping @Sizurka authorship of the main file.

**Comment by benpicco** [comment]
rebased to include vendor file updates.

File: cpu/sam0_common/include/exti_config.h [hunk 1]
```diff
@@ -0,0 +1,705 @@
...
+extern "C" {
+#endif
+
+/*
+ * Used in the ASF headers, but not always defined by them (looks like they
+ * sometimes define __L instead).
+ */
+#ifndef _L
+#define _L(i) i ## L
+#endif
```
Comment (dylad) [inline cpu/sam0_common/include/exti_config.h]: Is it really needed ?

Comment (benpicco) [reply cpu/sam0_common/include/exti_config.h]: Nope, looks like with the updated vendor files this is not needed anymore.

**Comment by benpicco** [comment]
saml21 & samr30 still need a vendor file update

**Comment by benpicco** [comment]
If I'm using `objdump` right, I see no change in the generated `exti_config` for SAMR30 compared to master:

`arm-none-eabi-objdump -d -j .text bin/samr30-xpro/default.elf | grep exti_config -A 3`

**master**
```
000088e8 <exti_config>:
    88e8:	0100 ffff 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    88f8:	0100 0302 ff04 0706 0d0c 0fff ff08 0b0a     ................
    8908:	ff00 0302 ffff ffff 0908 ffff ffff 0f0e     ................
```

**this branch**
```
000088d4 <exti_config>:
    88d4:	0100 ffff 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    88e4:	0100 0302 ff04 0706 0d0c 0fff ff08 0b0a     ................
    88f4:	ff00 0302 ffff ffff ffff ffff ffff 0fff     ................
```

**Comment by dylad** [comment]
@benpicco 
Something is wrong:
Your branch
```
$ arm-none-eabi-objdump -d -j .text bin/saml21-xpro/tests_periph_gpio.elf | grep exti_config -A 3
00003418 <exti_config>:
    3418:	ffffffff ffffffff ffffffff ffffffff     ................
    3428:	ffffffff ffffffff ffffffff ffffffff     ................
    3438:	ffffffff ffffffff ffffffff ffffffff     ................

```

master

$ arm-none-eabi-objdump -d -j .text bin/saml21-xpro/tests_periph_gpio.elf | grep exti_config -A 3
00003408 <exti_config>:
    3408:	03020100 07060504 0b0a09ff 0f0e0d0c     ................
    3418:	03020100 07060504 0fff0d0c 0b0affff     ................
    3428:	03020100 07060504 0b0a0908 0f0e0d0c     ................

**Comment by benpicco** [comment]
Uh, is your branch up to date?
I've rebased & force-pushed quite a bit - because I'm getting

```
% arm-none-eabi-objdump -d -j .text bin/saml21-xpro/tests_periph_gpio.elf| grep exti_config -A 3
00004144 <exti_config>:
    4144:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    4154:	0100 0302 0504 0706 0d0c 0fff ffff 0b0a     ................
    4164:	0100 0302 0504 0706 0908 0b0a 0d0c 0f0e     ................
```

**Comment by benpicco** [comment]
I've build `tests/periph_gpio` for all sam0 boards with this branch and `exti_config` is not empty:

<details>
<summary>small patch</summary>

```patch
--- a/cpu/sam0_common/Makefile.features
+++ b/cpu/sam0_common/Makefile.features
@@ -7,4 +7,6 @@ FEATURES_PROVIDED += periph_uart_modecfg
 FEATURES_PROVIDED += periph_uart_nonblocking
 FEATURES_PROVIDED += periph_wdt periph_wdt_cb
 
+FEATURES_PROVIDED += mcu_family_sam0
+
 -include $(RIOTCPU)/cortexm_common/Makefile.features
diff --git a/tests/periph_gpio/Makefile b/tests/periph_gpio/Makefile
index a99b20a222..1efc1d970f 100644
--- a/tests/periph_gpio/Makefile
+++ b/tests/periph_gpio/Makefile
@@ -3,6 +3,8 @@ include ../Makefile.tests_common
 FEATURES_REQUIRED = periph_gpio
 FEATURES_OPTIONAL = periph_gpio_irq
 
+FEATURES_REQUIRED += mcu_family_sam0
+
 USEMODULE += shell
 USEMODULE += benchmark
```
</details>

```bash
for i in $(make info-boards-supported); do
    make -j BOARD=$i && arm-none-eabi-objdump -d -j .text bin/$i/tests_periph_gpio.elf | grep exti_config -A 3 > $i.exti.log;
done
```

<details>
<summary>tail -n +1 *.exti.log</summary>

```
==> arduino-mkr1000.exti.log <==
0000618c <exti_config>:
    618c:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    619c:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    61ac:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> arduino-mkrfox1200.exti.log <==
0000618c <exti_config>:
    618c:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    619c:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    61ac:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> arduino-mkrwan1300.exti.log <==
00006230 <exti_config>:
    6230:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    6240:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    6250:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> arduino-mkrzero.exti.log <==
0000618c <exti_config>:
    618c:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    619c:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    61ac:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> arduino-zero.exti.log <==
00004234 <exti_config>:
    4234:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    4244:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    4254:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> feather-m0.exti.log <==
00006190 <exti_config>:
    6190:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    61a0:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    61b0:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> hamilton.exti.log <==
0000415f <exti_config>:
    415f:	ffff ffff ffff 0706 09ff 0b0a ffff 0f0e     ................
    416f:	0100 0302 ff04 ffff 0d0c 0fff ff08 0b0a     ................
    417f:	ff00 ffff ffff ffff 0908 ffff ffff 0f0e     ................

==> mlpa_tracker.exti.log <==
0000463c <exti_config>:
    463c:	03020100 07060504 0b0a09ff 0f0e0d0c     ................
    464c:	03020100 07060504 0bff0908 0f0effff     ................
    465c:	03020100 07060504 0b0a0908 0f0e0d0c     ................

==> mlpa_tracker-v2.exti.log <==
00004724 <exti_config>:
    4724:	03020100 07060504 0b0a09ff 0f0e0d0c     ................
    4734:	03020100 07060504 0bff0908 0f0effff     ................
    4744:	03020100 07060504 0b0a0908 0f0e0d0c     ................

==> samd21-xpro.exti.log <==
0000426c <exti_config>:
    426c:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    427c:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    428c:	0100 0302 0504 0706 0908 0b0a 0d0c 0f0e     ................

==> same54-xpro.exti.log <==
00004548 <exti_config>:
    4548:	03020100 07060504 0b0a09ff 0f0e0d0c     ................
    4558:	03020100 07060504 0bff0908 0f0effff     ................
    4568:	03020100 07060504 0b0a0908 0f0e0d0c     ................

==> saml10-xpro.exti.log <==
00003fcc <exti_config>:
    3fcc:	0100 0302 0504 0706 00ff 0201 ffff 0403     ................
    3fdc:	0605 0007 ffff 0201 0403 05ff ffff 0706     ................
    3fec:	7375 6761 3a65 2520 2073 703c 726f 3e74     usage: %s <port>

==> saml11-xpro.exti.log <==
00003fcc <exti_config>:
    3fcc:	0100 0302 0504 0706 00ff 0201 ffff 0403     ................
    3fdc:	0605 0007 ffff 0201 0403 05ff ffff 0706     ................
    3fec:	7375 6761 3a65 2520 2073 703c 726f 3e74     usage: %s <port>

==> saml21-xpro.exti.log <==
00004144 <exti_config>:
    4144:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    4154:	0100 0302 0504 0706 0d0c 0fff ffff 0b0a     ................
    4164:	0100 0302 0504 0706 0908 0b0a 0d0c 0f0e     ................

==> samr21-xpro.exti.log <==
00004294 <exti_config>:
    4294:	0100 ffff 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    42a4:	0100 0302 ff04 0706 0d0c 0fff ff08 0b0a     ................
    42b4:	ff00 0302 ffff ffff 0908 ffff ffff 0f0e     ................

==> samr30-xpro.exti.log <==
000040c4 <exti_config>:
    40c4:	0100 ffff 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    40d4:	0100 0302 ff04 0706 0d0c 0fff ff08 0b0a     ................
    40e4:	ff00 0302 ffff ffff 0908 ffff ffff 0f0e     ................

==> samr34-xpro.exti.log <==
000040c4 <exti_config>:
    40c4:	0100 ffff 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    40d4:	0100 0302 ffff 0706 0d0c 0fff ff08 0b0a     ................
    40e4:	ff00 0302 ffff ffff ffff ffff ffff 0fff     ................

==> sensebox_samd21.exti.log <==
00006298 <exti_config>:
    6298:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    62a8:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    62b8:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> sodaq-autonomo.exti.log <==
0000626c <exti_config>:
    626c:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    627c:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    628c:	0100 0302 0504 0706 0908 0b0a 0d0c 0f0e     ................

==> sodaq-explorer.exti.log <==
0000624c <exti_config>:
    624c:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    625c:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    626c:	0100 0302 0504 0706 0908 0b0a 0d0c 0f0e     ................

==> sodaq-one.exti.log <==
00006288 <exti_config>:
    6288:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    6298:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    62a8:	ffff 0302 ffff ffff 0908 0b0a ffff ffff     ................

==> sodaq-sara-aff.exti.log <==
000062ac <exti_config>:
    62ac:	0100 0302 0504 0706 09ff 0b0a 0d0c 0f0e     ................
    62bc:	0100 0302 0504 0706 0d0c 0fff ff08 0b0a     ................
    62cc:	0100 0302 0504 0706 0908 0b0a 0d0c 0f0e     ................
```
</details>

**Comment by dylad** [comment]
@benpicco 
We should be good.
Please add a small notice about the vendor files modification in the dedicated README. You can squash directly.

**Comment by benpicco** [comment]
I added a note to `vendor/README.md`.

**Comment by benpicco** [comment]
Now we only need that device name parsing like in stm32 ;)
But I'm about to add a samd21e18a board, so this comes handy.

**Pull request merged**